include/sys/queued_operation.h

```diff
@@ -0,0 +1,343 @@
+	 * should itself be async, with finalization done after
+	 * process() returns.
+	 *
+	 * @note The process function must be isr-ok.
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation that should be initiated.  A null
+	 * pointer is passed if there are no pending operations.
+	 */
+	void (*process)(struct queued_operation_manager *mgr,
+			struct queued_operation *op);
+};
+
+/**
+ * @brief State associated with a manager instance.
+ */
+struct queued_operation_manager {
+	/* Links the operation into the operation queue. */
+	sys_slist_t operations;
+
+	/* Pointer to the functions that support the manager. */
+	const struct queued_operation_functions *vtable;
+
+	/* Pointer to an on-off service supporting this service.  NULL
+	 * if service is always available.
+	 */
+	struct onoff_manager *onoff;
+
+	/* The state of on-off service requests. */
+	struct onoff_client onoff_client;
+
+	/* Lock controlling access to other fields. */
+	struct k_spinlock lock;
+
+	/* The operation that is being processed. */
+	struct queued_operation *current;
+
+	/* Information about the internal state of the manager. */
+	uint32_t volatile state;
+};
+
+#define QUEUED_OPERATION_MANAGER_INITIALIZER(_vtable, _onoff) {	\
+		.vtable = _vtable,				\
+		.onoff = _onoff,				\
+}
+
+/**
+ * @brief Submit an operation to be processed when the service is
+ * available.
+ *
+ * The service process function will be invoked during this call if
+ * the service is available.
+ *
+ * @param mgr a generic pointer to the service instance
+ *
+ * @param op a generic pointer to an operation to be performed.  The
+ * notify field in the provided operation must have been initialized
+ * before being submitted, even if the operation description is being
+ * re-used.  This may be done directly with sys_notify API or by
+ * wrapping it in a service-specific operation init function.
+ *
+ * @param priority the priority of the operation relative to other
+ * operations.  Numerically lower values are higher priority.  Values
+ * outside the range of a signed 8-bit integer will be rejected,
+ * except for named priorities like QUEUED_OPERATION_PRIORITY_APPEND.
+ *
+ * @retval -ENOTSUP if callback notification is requested and the
+ * service does not provide a callback translation.  This may also be
+ * returned due to service-specific validation.
+ *
+ * @retval -EINVAL if the passed priority is out of the range of
+ * supported priorities.  This may also be returned due to
+ * service-specific validation.
+ *
+ * @return A negative value if the operation was rejected by service
+ * validation or due to other configuration errors.  A non-negative
+ * value indicates the operation has been accepted for processing and
+ * completion notification will be provided.
+ */
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority);
+
```

Comment:
what about adding helper function like
```
int queued_operation_sync_submit(struct queued_operation_manager *mgr,
           struct queued_operation *op,  int priority);
```
which from thread context would schedule operation and pend for completion. That would safe complexity of setting up `sys_notify`. As i have some work based on this branch I've added there something like this which is relying on `POLL` being enabled (so semaphore version would need to be added as well):
```
int queued_operation_sync_submit(struct queued_operation_manager *qop_mgr,
			struct queued_operation *qop, int priority)
{
	int err;
	int ret;
	struct k_poll_signal signal = K_POLL_SIGNAL_INITIALIZER(signal);
	struct k_poll_event event = K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,
							K_POLL_MODE_NOTIFY_ONLY,
							&signal);

	sys_notify_init_signal(&qop->notify, &signal);
	err = queued_operation_submit(qop_mgr, qop, priority);
	if (err < 0) {
		return err;
	}

	err = k_poll(&event, 1, K_FOREVER);
	if (err < 0) {
		return err;
	}

	err = sys_notify_fetch_result(&qop->notify, &ret);
	if (err < 0) {
		return err;
	}

	return ret;
}
```

Reply:
Like the other resource management APIs this really isn't intended for direct use by application code.  At a minimum there should be a service-specific API that can do type checking on the operation before it's submitted.  In the reference code in the test that is `service_submit()`.  I would expect there to be a companion `service_submit_sync()` to provide this feature.

I don't think lifting a generic `submit_sync` API into the manager is worthwhile at this time.  It's possible that may change in the future, but locking it down when there are no in-tree users of this functionality is premature.